microceph/ceph/osd.go

```diff
@@ -438,28 +439,407 @@ func AddOSD(s *state.State, path string, wipe bool, encrypt bool) error {
+// if we remove the given OSD
+func IsDowngradeNeeded(s common.StateInterface, osd int64) (bool, error) {
+	currentRule, err := getDefaultCrushRule()
+	if err != nil {
+		return false, err
+	}
+	hostRule, err := getCrushRuleID("microceph_auto_host")
+	if err != nil {
+		return false, err
+	}
+	if currentRule != hostRule {
+		// either we're at 'osd' level or we're using a custom rule
+		// in both cases we won't downgrade
+		logger.Infof("No need to downgrade auto failure domain, current rule is %v", currentRule)
+		return false, nil
+	}
+	numNodes, err := database.MemberCounter.CountExclude(s.ClusterState(), osd)
+	logger.Infof("Number of nodes excluding osd.%v: %v", osd, numNodes)
+	if err != nil {
+		return false, err
+	}
+	if numNodes < 3 { // need to scale down
+		return true, nil
+	}
+	return false, nil
+}
+
+// scaleDownFailureDomain scales down the failure domain from 'host' to 'osd' level
+func scaleDownFailureDomain(s common.StateInterface, osd int64) error {
+	needDowngrade, err := IsDowngradeNeeded(s, osd)
+	logger.Debugf("Downgrade needed: %v", needDowngrade)
+	if err != nil {
+		return err
+	}
+	if !needDowngrade {
+		return nil
+	}
+	err = switchFailureDomain("host", "osd")
+	if err != nil {
+		return fmt.Errorf("Failed to switch failure domain: %w", err)
+	}
+	return nil
+}
+
+// reweightOSD reweights the given OSD to the given weight
+func reweightOSD(ctx context.Context, osd int64, weight float64) {
+	logger.Debugf("Reweighting osd.%d to %f", osd, weight)
+	_, err := processExec.RunCommand(
+		"ceph", "osd", "crush", "reweight",
+		fmt.Sprintf("osd.%d", osd),
+		fmt.Sprintf("%f", weight),
+	)
+	if err != nil {
+		// only log a warn, don't treat fail to reweight as a fatal error
+		logger.Warnf("Failed to reweight osd.%d: %v", osd, err)
+	}
+}
+
+func doPurge(osd int64) error {
+	// run ceph osd purge command
+	_, err := processExec.RunCommand(
+		"ceph", "osd", "purge", fmt.Sprintf("osd.%d", osd),
+		"--yes-i-really-mean-it",
+	)
+	return err
+}
+
+func purgeOSD(osd int64) error {
+	var err error
+	retries := 10
+	var backoff time.Duration
+
+	for i := 0; i < retries; i++ {
+		err = doPurge(osd)
+		if err == nil {
+			// Success: break the retry loop
+			break
+		}
+		// we're getting a RunError from processExec.RunCommand, and it
+		// wraps the original exit error if there's one
+		exitError, ok := err.(shared.RunError).Unwrap().(*exec.ExitError)
+		if !ok {
+			// not an exit error, abort and bubble up the error
+			logger.Warnf("Purge failed with non-exit error: %v", err)
+			break
+		}
+		if syscall.Errno(exitError.ExitCode()) != syscall.EBUSY {
+			// not a busy error, abort and bubble up the error
+			logger.Warnf("Purge failed with unexpected exit error: %v", exitError)
+			break
+		}
+		// purge failed with EBUSY - retry after a delay, and make delay exponential
+		logger.Infof("Purge failed %v, retrying in %v", err, backoff)
+		backoff = time.Duration(math.Pow(2, float64(i))) * time.Millisecond * 100
```

Comment:
It's a nitpick, but I frankly don't see an exponential backoff as something useful here. This whole function is called under a timeout, so the backoff can actually lead to some retries not being attempted. I would use a fixed-time sleep, but it's just a personal preference.

Reply:
Not a must-have IMO but I'd lean towards keeping it in as on a busy system it can give the cluster a bit more breathing room and keeps the logs somewhat cleaner (and it doesn't cost much to implement). Wrt to the timeout, this is user-configurable, so if the timeout kicks in we don't control